Attempt to implement  visualization

Zavioer opened the pull request:
Basic visualization based on plotly/dash framework.

Zavioer committed:
Attempt toi mplement  visualization

Basic visualization based on plotly/dash framework.

Zavioer committed:
Implement simulation set up and control

Build callback method for set up and control (start, stop, reset) simulation in dash framework.

Zavioer committed:
Simulation page base on dash update

Update webpage layout and add handlin q and R0 parameters.

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+        r_parameter = -change_susceptible / (data['groups']['susceptible'] * data['groups']['infected'])
+
+    if data['groups']['infected'] > 0 and data['groups']['dead'] > 0 and data['groups']['recovered'] > 0:
+        a_parameter = change_removed * (1 / data['groups']['infected'] * (data['groups']['dead'] +
```
Wrong equation. Look at: https://github.com/amasend/SIR-simulation-IBM-ESI/issues/18

amasend commented on requirements.txt:
```diff
@@ -1 +1,4 @@
-
+# Core modules #
+plotly
+dash
+pandas
```
It is always better to specify versions.

amasend commented on tests/test_interaction.py:
```diff
@@ -1,7 +1,5 @@
 import unittest
 from simulation import person
-from simulation import container
-import plotly.graph_objects as go
 
 
 class InfectedAndSusceptibleInteractionTestCase(unittest.TestCase):
```
Where are tests for each big method from `app.py`? I am not talking about plots etc. but rather about functionality, `R0` `q` etc.

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    change_d = abs(old_amounts['dead'] - data['groups']['dead'])
+    change_r = abs(old_amounts['recovered'] - data['groups']['recovered'])
+    change_removed = change_d + change_r
+    r_parameter = 0
+    a_parameter = 1
+    q_parameter = 0
+    r_0 = 0
```
Is there really a need to keep these variables every time redefined?

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    """
+
+    data = json.loads(json_data)
+    change_susceptible = data['groups']['susceptible'] - old_amounts['susceptible']
+    change_d = abs(old_amounts['dead'] - data['groups']['dead'])
+    change_r = abs(old_amounts['recovered'] - data['groups']['recovered'])
+    change_removed = change_d + change_r
```
Maybe better to place such functionality in separate functions? To somehow segragate, for better visibility of functionality.

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    Html table base od bash html.Table object.
+    """
+
+    data = json.loads(json_data)
```
Why this is a binary data? Why not a normal python dict? Each time you need to load it and parse it. Is there any reason?

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    ----------
+    n_intervals: int, required
+        Number of past intervals.
+    json_data: json, required
```
`json` is not a type of an object here, `json` is a text representation for computer and humans. Better to use dictionary.

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    """
+
+    data = json.loads(json_data)
+    percent = {'condition': [], 'percent_value': []}
+
+    percent['condition'].append('susceptible')
+    percent['percent_value'].append(data['groups']['susceptible'])
+    percent['condition'].append('infected')
+    percent['percent_value'].append(data['groups']['infected'])
+    percent['condition'].append('recovered')
+    percent['percent_value'].append(data['groups']['recovered'])
+    percent['condition'].append('dead')
+    percent['percent_value'].append(data['groups']['dead'])
```
If you create `percent` per each iteration, it will be faster to place whole data in one place at one time instead of appending lists.

```suggestion
    percent = {'condition': ['susceptible', 'infected' ...], 'percent_value': [data['groups']['susceptible'], data['groups']['infected'] ...]}
```

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+
+    data = json.loads(json_data)
+
+    d['time'].append(n_intervals)
+    d['value'].append(data['groups']['susceptible'])
+    d['condition'].append('susceptible')
+    d['time'].append(n_intervals)
+    d['value'].append(data['groups']['infected'])
+    d['condition'].append('infected')
+    d['time'].append(n_intervals)
+    d['value'].append(data['groups']['recovered'])
+    d['condition'].append('recovered')
+    d['time'].append(n_intervals)
+    d['value'].append(data['groups']['dead'])
+    d['condition'].append('dead')
```
Maybe there is a method to preserve last content of the graph ans draw only next points? I do not know, maybe there is something better than appending each time and draw entire plot from scratch. Need to search.

amasend commented on simulation/app.py:
```diff
@@ -0,0 +1,621 @@
+    ])
+
+
+@app.callback(Output('inf-prob-value', 'children'),
+              [Input('infection-probability', 'value')])
+def update_slider(value):
+    return value
+
+
+@app.callback(Output('rec-prob-value', 'children'),
+              [Input('recover-probability', 'value')])
+def update_slider(value):
+    return value
+
+
+@app.callback(Output('dead-prob-value', 'children'),
+              [Input('dead-probability', 'value')])
+def update_slider(value):
+    return value
+
+
+@app.callback(Output('inf-range-value', 'children'),
+              [Input('infection-range', 'value')])
+def update_slider(value):
+    return value
+
+
+@app.callback(Output('move-dist-value', 'children'),
+              [Input('move-length', 'value')])
+def update_slider(value):
+    return value
```
Here you have redeclaration of functions even without description. Ofc. it is better here to name function individually than putting docstrings as these functions are really small.

amasend reviewed (changes requested):
**UI:**
Placement of S I R D numbers and R0 / q could be placed maybe on the right side of the main plot? Right now it is somehow weird, too much space there.

Maybe a good idea is to calculate and place the mean values of `R0` and `q` through a time? I mean along with instantaneous values like it is now. Mean value could expose the real expected values better than some peaks like now.

**Code:**
Still missing testing part of some components.

Overall really good. It is going to better direction ;)

Zavioer committed:
Add unitests and change dashboard layout

Add unitests for functions used in app.py and modified bashboard layout with tables on right side of main plot.

Zavioer committed:
Change visualisation of R0 and q

Update visualisation of R0 and q parameter, add input field to specified container size. Change marker type of infected instances in move chart.

Zavioer committed:
Update main plot

Separate to additional file functions for computing r, a, q and R0 parameters. Rebuild main plot with plolty.graph_objects, add to plot R0 parameter.

Zavioer committed:
Update syntax mistakes

Zavioer committed:
Update app.py
